core/src/apps/ethereum/sign_tx.py

```diff
@@ -208,3 +292,86 @@ def check_common_fields(msg: MsgInSignTx) -> None:
 
     if msg.chain_id == 0:
         raise DataError("Chain ID out of bounds")
+
+
+async def _handle_staking_tx_stake(
+    data_reader: BufferReader,
+    msg: MsgInSignTx,
+    network: EthereumNetworkInfo,
+    to_address: bytes,
+    maximum_fee: str,
+    fee_items: Iterable[tuple[str, str]],
+) -> None:
+    from .layout import require_confirm_stake
+
+    # stake args:
+    # - arg0: uint64, source (should be 1)
+    if data_reader.remaining_count() != constants.SC_ARGUMENT_BYTES:
+        raise DataError  # wrong number of arguments for stake (should be 1)
+    if (
+        int.from_bytes(data_reader.read_memoryview(constants.SC_ARGUMENT_BYTES), "big")
+        != 1
+    ):
+        raise DataError  # wrong value of 1st argument ('source' should be 1)
+
+    await require_confirm_stake(
+        to_address,
+        int.from_bytes(msg.value, "big"),
+        maximum_fee,
+        fee_items,
+        network,
+        bool(msg.chunkify),
+    )
+
+
+async def _handle_staking_tx_unstake(
+    data_reader: BufferReader,
+    msg: MsgInSignTx,
+    network: EthereumNetworkInfo,
+    to_address: bytes,
+    maximum_fee: str,
+    fee_items: Iterable[tuple[str, str]],
+) -> None:
+    from .layout import require_confirm_unstake
+
+    # unstake args:
+    # - arg0: uint256, value
+    # - arg1:  uint16, isAllowedInterchange (bool)
+    # - arg2: uint64, source, should be 1
+    if data_reader.remaining_count() != constants.SC_ARGUMENT_BYTES * 3:
+        raise DataError  # wrong number of arguments for unstake (should be 3)
+    value = int.from_bytes(
+        data_reader.read_memoryview(constants.SC_ARGUMENT_BYTES), "big"
+    )
+    _ = data_reader.read_memoryview(constants.SC_ARGUMENT_BYTES)  # skip arg1
+    if (
+        int.from_bytes(data_reader.read_memoryview(constants.SC_ARGUMENT_BYTES), "big")
+        != 1
+    ):
+        raise DataError  # wrong value of 3rd argument ('source' should be 1)
+
+    await require_confirm_unstake(
+        to_address,
+        value,
+        maximum_fee,
+        fee_items,
+        network,
+        bool(msg.chunkify),
+    )
+
+
+async def _handle_staking_tx_claim(
+    data_reader: BufferReader,
+    staking_addr: bytes,
+    maximum_fee: str,
+    fee_items: Iterable[tuple[str, str]],
+    network: EthereumNetworkInfo,
+    chunkify: bool,
+) -> None:
+    from .layout import require_confirm_claim
+
+    # claim has no args
+    if data_reader.remaining_count() != 0:
+        raise DataError  # wrong number of arguments for claim (should be 0)
```

Comment:
always provide text for the DataError. (you can use the same one across all functions, something like "Invalid smart contract call" seems appropriate)
DataError strings are not supposed to be translated (for now)

Reply:
Ok, provided message for `DataError` https://github.com/trezor/trezor-firmware/pull/3534/commits/339da2725da19a08e116a30fa035fcb17f31a463